src/System.Linq.Expressions/src/System/Linq/Expressions/Interpreter/TypeOperations.cs

```diff
@@ -2210,44 +2210,85 @@ private class CastInstructionT<T> : CastInstruction
             public override int Run(InterpretedFrame frame)
             {
                 var value = frame.Pop();
-                if (value != null)
-                {
-                    frame.Push((T)value);
-                }
-                else
-                {
-                    frame.Push(null);
-                }
+                frame.Push((T)value);
                 return +1;
             }
         }
 
-        private class CastInstructionNoT : CastInstruction
+        private abstract class CastInstructionNoT : CastInstruction
         {
             private readonly Type _t;
-            public CastInstructionNoT(Type t)
+            protected CastInstructionNoT(Type t)
             {
                 _t = t;
             }
 
+            public new static CastInstruction Create(Type t)
+            {
+                if (t.GetTypeInfo().IsValueType && !TypeUtils.IsNullableType(t))
+                {
+                    return new Value(t);
+                }
+                else
+                {
+                    return new Ref(t);
+                }
+            }
+
             public override int Run(InterpretedFrame frame)
             {
                 var value = frame.Pop();
                 if (value != null)
                 {
-                    if (!TypeUtils.HasReferenceConversion(value.GetType(), _t) &&
-                        !TypeUtils.HasIdentityPrimitiveOrNullableConversion(value.GetType(), _t))
+                    var valueType = value.GetType();
+
+                    if (!TypeUtils.HasReferenceConversion(valueType, _t) &&
+                        !TypeUtils.HasIdentityPrimitiveOrNullableConversion(valueType, _t))
+                    {
+                        throw new InvalidCastException();
+                    }
+
+                    if (!_t.IsAssignableFrom(valueType))
```

Comment:
I'll leave the old conditions in here and do some more analysis to see whether any of it is redundant due to the addition of `IsAssignableFrom`. Seems to me this PR is ready to go for now, can you have a second look and sign off / pull?